### PR TITLE
support create group

### PR DESCRIPTION
**Pull request by imranariffin**
Resolves #40 

* use transaction to create message group & add message group member atomically
* collect incoming action validation error before responding
* log error before returning

### CODE REVIEW

**Comment by imranariffin** [comment]
@cglotr Can you review this?

**Comment by cglotr** [comment]
> @cglotr Can you review this?

@imranariffin Will do it today. Busy with other stuff yesterday ahaha

**Comment by imranariffin** [comment]
> > @cglotr Can you review this?
> 
> @imranariffin Will do it today. Busy with other stuff yesterday ahaha

No hurry, just thought you forgot about it haha.

File: db.go [hunk 1]
```diff
@@ -372,3 +373,84 @@ func updateReceiptDateDelivered(messageID, recipientID int) error {
 	`, messageID, recipientID)
 	return err
 }
+
+func createMessageGroup(groupName string, memberIds []int) (*MessageGroup, error) {
+	tx, err := db.Begin()
+
+	if err != nil {
```
Comment (cglotr) [inline db.go:380]: I kinda have a new convention that is _Golang_ specific about `if` spacing. Usually, I put a newline before & after `if`, but for _Golang_ I feel that _err_ or _ok_ `if`s should be in the same block as the declaration statement. Something like below:
```
	tx, err := db.Begin()
	if err != nil {
		log.Println(err)
		return nil, err
	}
```

File: db.go [hunk 2]
```diff
@@ -372,3 +373,84 @@ func updateReceiptDateDelivered(messageID, recipientID int) error {
...
+		log.Println(err)
+		return nil, err
+	}
+
+	result, err := tx.Exec(
+		`INSERT INTO message_group ( name ) VALUES ( ? );`,
```
Comment (cglotr) [inline db.go:386]: Kinda like the spacing inside `( xxx )`. It looks more readable and nicer. Let's apply it across all SQL statements?

File: constants.go [hunk 3]
```diff
@@ -5,6 +5,10 @@ const (
 	authorizationSignInRequest = "AUTHORIZATION_SIGN_IN_REQUEST"
 	authorizationSignInSuccess = "AUTHORIZATION_SIGN_IN_SUCCESS"
 
+	groupCreateFailure = "GROUP_CREATE_FAILURE"
```
Comment (cglotr) [inline constants.go:8]: I think the name should be `MESSAGE_GROUP_CREATE_*` since the table is named as such.

File: db.go [hunk 4]
```diff
@@ -372,3 +373,84 @@ func updateReceiptDateDelivered(messageID, recipientID int) error {
...
+			)
+			log.Println(err)
+			return nil, err
+		}
+	}
+
+	err = tx.Commit()
+
+	if err != nil {
+		log.Fatal(err)
```
Comment (cglotr) [inline db.go]: `log.Fatal` will terminate the server! From the [docs](https://golang.org/pkg/log/#Fatal): _"Fatal is equivalent to Print() followed by a call to os.Exit(1)."_

File: db.go [hunk 5]
```diff
@@ -372,3 +373,84 @@ func updateReceiptDateDelivered(messageID, recipientID int) error {
...
+	groupId, err := result.LastInsertId()
+
+	if err != nil {
+		tx.Rollback()
+		log.Println(err)
+		return nil, err
+	}
+
+	for _, memberId := range memberIds {
+		result, err = tx.Exec(
```
Comment (cglotr) [inline db.go:405]: I prefer to write it like this:
```
	result, err = tx.Exec(
		`
			INSERT INTO message_group_member ( message_group_id, member_id )
			VALUES ( ?, ? );
		`,
		groupId,
		memberId,
	)
```
I consider ` to be like an opening & closing bracket so I write them as such. Not sure if it makes sense to other people tho ahaha.

File: structs.go [hunk 6]
```diff
@@ -36,3 +36,11 @@ type MessageDelivery struct {
 	Message      *Message
 	RecipientIDs []int
 }
+
+// MessageGroup .
+type MessageGroup struct {
```
Comment (cglotr) [inline structs.go:41]: The columns should be sorted as in the DB. And I think `MemberIDs` should be put after a newline to indicate that it's different with the rest of the columns.

File: v2.go [hunk 7]
```diff
@@ -333,3 +335,59 @@ func handleVerificationSubmitCodeRequest(conn *websocket.Conn, action *Action) {
 	verifyClient(client, conn, code)
 	writeEmptyAction(conn, verificationSubmitCodeSuccess)
 }
+
+func handleGroupCreateRequest(conn *websocket.Conn, action *Action) {
+
```
Comment (cglotr) [inline v2.go:340]: Rogue newline.

File: v2.go [hunk 8]
```diff
@@ -333,3 +335,59 @@ func handleVerificationSubmitCodeRequest(conn *websocket.Conn, action *Action) {
...
+	client, err := getClient(conn)
+
+	if err != nil {
+		errors = append(errors, err.Error())
+		writeFailure(conn, groupCreateFailure, errors)
+		return
+	}
+
+	groupName, groupNameOk := action.Payload["name"].(string)
+	_memberIds, memberIdsOk := action.Payload["member_ids"].([]interface{})
```
Comment (cglotr) [inline v2.go:351]: Not sure if we should introduce `_xxx` variable name in _Golang_. I find it to affect readability at times.

File: db.go [hunk 9]
```diff
@@ -372,3 +373,84 @@ func updateReceiptDateDelivered(messageID, recipientID int) error {
 	`, messageID, recipientID)
 	return err
 }
+
+func createMessageGroup(groupName string, memberIds []int) (*MessageGroup, error) {
+	tx, err := db.Begin()
```
Comment (cglotr) [inline db.go:378]: Good use of _transaction_ I think.

By the way, should we make _transaction_ a default way of doing SQL? Are there any downsides to it?

**Review by cglotr — CHANGES_REQUESTED** [review]
There are a few minor _syntax_ related comments among other things. LGTM otherwise.

I have added one more commit (bb9f4ff7b3d7e22c866b881055942662cc8bb58e) to fix message delivery bug since it's a bug from my feature.

File: constants.go (hunk 3 above)
Comment (cglotr) [reply constants.go:8]: Actually, I think I agree with the name. It shouldn't be tied to table name

File: db.go (hunk 2 above)
Comment (cglotr) [reply db.go:386]: Done f30ffa5

File: db.go (hunk 4 above)
Comment (cglotr) [reply db.go]: Done 48113bc

**Review by cglotr — APPROVED** [review]
LGTM. Merging

File: db.go (hunk 9 above)
Comment (imranariffin) [reply db.go:378]: I think we should use it only when needed. Not sure about downsides though.

File: db.go (hunk 5 above)
Comment (imranariffin) [reply db.go:405]: It makes the sql look more obvious but may also look redundant. But it definitely uses more indentation.

It's hard for me to get on board with this idea.

File: v2.go (hunk 8 above)
Comment (imranariffin) [reply v2.go:351]: @cglotr
Actually I did wonder if _xxx is within standard golang convention but forgot to look it up.

If it is not within the standard convention, do you have any idea on how to rename it or maybe a different if doing this that does not need such temporary variable?

**Pull request merged**